src/common/s3.connection.cpp

```diff
@@ -0,0 +1,569 @@
+    TRACE("Creating bucket %s", bucket_name.c_str());
+    minio::s3::MakeBucketArgs args;
+    args.bucket = bucket_name;
+
+    try {
+        minio::s3::MakeBucketResponse response = client_->MakeBucket(args);
+
+        return (bool)response;
+    } catch (const std::exception& e) {
+        LOGE("Failed to create bucket %s: %s", bucket_name.c_str(), e.what());
+    }
+
+    return false;
+}
+
+bool
+common::S3Connection::destroy_bucket(const std::string& bucket_name)
+{
+    if (bucket_name.empty()) {
+        return false;
+    }
+
+    // first check if the bucket exists, do nothing if it does
+    if (!bucket_exists(bucket_name)) {
+        return true;
+    }
+
+    TRACE("Destroying bucket %s", bucket_name.c_str());
+    minio::s3::RemoveBucketArgs args;
+    args.bucket = bucket_name;
+
+    try {
+        minio::s3::RemoveBucketResponse response = client_->RemoveBucket(args);
+
+        return (bool)response;
+    } catch (const std::exception& e) {
+        LOGE("Failed to destroy bucket %s: %s", bucket_name.c_str(), e.what());
+    }
+
+    return false;
+}
+
+bool
+common::S3Connection::object_exists(const std::string& bucket_name,
+                                    const std::string& object_name)
+{
+    if (bucket_name.empty() || object_name.empty()) {
+        return false;
+    }
+
+    try {
+        minio::s3::StatObjectArgs args;
+        args.bucket = bucket_name;
+        args.object = object_name;
+
+        minio::s3::StatObjectResponse response = client_->StatObject(args);
+
+        return (bool)response;
+    } catch (const std::exception& e) {
+        LOGE("Failed to check if object %s exists in bucket %s: %s",
+             object_name.c_str(),
+             bucket_name.c_str(),
+             e.what());
+    }
+
+    return false;
+}
+
+bool
+common::S3Connection::put_object(const std::string& bucket_name,
+                                 const std::string& object_name,
+                                 const uint8_t* data,
+                                 size_t nbytes,
+                                 std::string& etag)
+{
+    if (bucket_name.empty() || object_name.empty() || 0 == nbytes) {
+        return false;
+    }
+
+    std::stringstream ss;
+    std::copy(reinterpret_cast<const char*>(data),
+              reinterpret_cast<const char*>(data) + nbytes,
+              std::ostream_iterator<char>(ss));
+
+    TRACE("Putting object %s in bucket %s",
+          object_name.c_str(),
+          bucket_name.c_str());
+    minio::s3::PutObjectArgs args(ss, (long)nbytes, 0);
+    args.bucket = bucket_name;
+    args.object = object_name;
+
+    try {
+        minio::s3::PutObjectResponse response = client_->PutObject(args);
+        etag = response.etag;
+
+        return (bool)response;
+    } catch (const std::exception& e) {
+        LOGE("Failed to put object %s in bucket %s: %s",
+             object_name.c_str(),
+             bucket_name.c_str(),
+             e.what());
+    }
+
+    return false;
+}
```

Comment:
> Instead of using std::stringstream, consider passing a pointer and size directly if the API supports it. This can improve performance.

I wish it did. This is my convoluted attempt to get it working. The [CharBuffer](https://github.com/minio/minio-cpp/blob/cd4ef14955201d5ce81e54155ef0980f7c644864/include/miniocpp/utils.h#L221) class might work, as I mentioned in another comment. I'm going to try that.